test/unit/lib/plugins/aws/deploy/index.test.js

```diff
@@ -348,3 +220,244 @@ describe('AwsDeploy', () => {
     });
   });
 });
+
+describe('test/unit/lib/plugins/aws/deploy/index.test.js', () => {
+  const baseAwsRequestStubMap = {
+    STS: {
+      getCallerIdentity: {
+        ResponseMetadata: { RequestId: 'ffffffff-ffff-ffff-ffff-ffffffffffff' },
+        UserId: 'XXXXXXXXXXXXXXXXXXXXX',
+        Account: '999999999999',
+        Arn: 'arn:aws:iam::999999999999:user/test',
+      },
+    },
+  };
+
+  it('with nonexistent stack - first deploy', async () => {
+    const describeStacksStub = sinon
+      .stub()
+      .onFirstCall()
+      .throws('error', 'stack does not exist')
+      .onSecondCall()
+      .resolves({ Stacks: [{}] });
+    const createStackStub = sinon.stub().resolves({});
+    const s3UploadStub = sinon.stub().resolves();
+    const describeStackResourceStub = sinon
+      .stub()
+      .resolves({ StackResourceDetail: { PhysicalResourceId: 's3-bucket-resource' } });
+    const validateTemplateStub = sinon.stub().resolves();
+    const updateStackStub = sinon.stub().resolves({});
+    const listObjectsV2Stub = sinon.stub().resolves({ Contents: [] });
+    const awsRequestStubMap = {
+      ...baseAwsRequestStubMap,
+      S3: {
+        listObjectsV2: listObjectsV2Stub,
+        upload: s3UploadStub,
+      },
+      CloudFormation: {
+        describeStacks: describeStacksStub,
+        createStack: createStackStub,
+        describeStackEvents: {
+          StackEvents: [
+            {
+              EventId: '1e2f3g4h',
+              StackName: 'new-service-dev',
+              LogicalResourceId: 'new-service-dev',
+              ResourceType: 'AWS::CloudFormation::Stack',
+              Timestamp: new Date(),
+              ResourceStatus: 'CREATE_COMPLETE',
+            },
+          ],
+        },
+        describeStackResource: describeStackResourceStub,
+        validateTemplate: validateTemplateStub,
+        updateStack: updateStackStub,
+        listStackResources: {},
+      },
+    };
+
+    const { awsNaming, cfTemplate } = await runServerless({
+      fixture: 'function',
+      cliArgs: ['deploy'],
+      awsRequestStubMap,
+    });
+
+    expect(createStackStub).to.be.calledOnce;
+    // Called with artifact and CloudFormation template
+    expect(s3UploadStub.callCount).to.equal(2);
+    expect(s3UploadStub.args[0][0].Key.endsWith('compiled-cloudformation-template.json')).to.be
+      .true;
+    expect(s3UploadStub.args[0][0].Body).to.equal(JSON.stringify(cfTemplate));
+    expect(s3UploadStub.args[1][0].Key.endsWith(awsNaming.getServiceArtifactName())).to.be.true;
+    expect(describeStackResourceStub).to.be.calledOnce;
+    expect(validateTemplateStub).to.be.calledOnce;
+    expect(updateStackStub).to.be.calledOnce;
+    // Second call as a part of cleanupS3Bucket
+    expect(listObjectsV2Stub).to.be.calledTwice;
```

Comment:
This is where I had really mixed feelings. I totally agree with assertion on `deleteObjects`, but what about testing scenario where we want to check that `cleanup` was triggered, but `deleteObjects` was not called because there was no artifacts to cleanup?

Reply:
> here we want to check that cleanup was triggered, but deleteObjects was not called because there was no artifacts to cleanup?

Why do you think it's worth testing that? I think it's only external effect that matters, and not that internally something was called or not.

I'd just confirm that if there are artifacts to cleanup they are cleanup as expected, and if there's nothing to cleanup, we may simply confirm that there was no attempt to delete objects that are expected to stay.

Reply:
I agree that it's not testing actual outcomes in this case, my idea here was to check if the logic is actually triggered in this scenario but you're right that there're no visible outcomes anyway. :+1:

Reply:
Yes, `listObjects` technically is also an external API, on which we could confirm. Still I'd rather test only those calls that introduce side effects, so ones that create and remove.

Theoretically how we resolved a list of objects to delete should be treated as black box. What's to be tested is that we pushed the right list of objects to destructive methods